### DISCUSSION TITLE
Design-by-contract

### DISCUSSION THREAD

**Discussion by davidkpiano** [statelyai/xstate#2322] [Ideas]
Let's talk about how we can bring ideas from [design-by-contract](https://en.wikipedia.org/wiki/Design_by_contract) into XState, which can potentially replace Typestates, which are helpful, but not robust (it's easy to defy). 

Some ideas from the [Discord](discord.gg/xstate):

https://discord.com/channels/795785288994652170/799416911259369472/850539410218811423

```ts
// TENTATIVE API
states: {
  // ...
  loaded: {
    // for all entering transitions
    pre: assertUserExists,
    // while in this state
    invariant: assertUserExists,
    // for all exiting transitions
    post: assertUserIsNull
  }
}
```

Can make use of this:

```ts
type Asserter<T> = (a: any) => asserts a is T;

type GetAssertValue<T extends Asserter<any>> = T extends Asserter<infer U> ? U : never;

declare function foo(bar: any): asserts bar is string;

type Foo = GetAssertValue<typeof foo>; // string
```

Tested out: https://codesandbox.io/s/typestates-hkku5?file=/src/createMachine2.ts:2871-3102

Addresses #1060

**Comment by devanshj** [comment]
In light of #2323 here's [txstate](https://github.com/devanshj/txstate)'s philosophy on typestates. Needless to say this is not to undermine the efforts by others.

Typestates can be fully inferred without needing the user to do anything special, nothing, zero. Here's how... In xstate the only way you can mutate the context is via the `assign` action, there's no other way (afaik). And moreover we know what mutations are being done on the type-level just by accessing the return type of the assign action. So we exactly know when, where and how the context is getting mutated by the machine - that's exactly what all we need to infer typestates.

Okay if it's not making any sense to you (xD) think about it like this - If I show you a machine will you, as a developer, not be able to derive the typestates looking at when, where and how the mutations to the context are being made? So why can't the types do it.

Moreover, if you allow users to write typestates (explicitly or via design-by-contract assertions) it's entirely possible they write it incorrectly. Meaning let's say with dbc the user writes `invariant: assertUserExists`, though they never assign `user` in the entry action, in that case the invariant is going to fail in runtime. If we infer (derive) typestates instead of having the user define them we can bring the invariant on the type level and it's going to be a compile error (when the user will write `context.user` somewhere) because they never assigned it in the entry. (I'm taking entry action as an example, sure there are many ways to do it, like having an action in the transition that targets the node)

My two cents, ignore if it seems I'm on crack but I'm not xD

Edit:
We can have design-by-contract assertions too, for the sake of runtime. But along with inferring typestates, now we'll be able to validate assertions themselves. So like if the user doesn't assign `user` in the entry action then they'll see red underlines on `invariant: assertUserExists` because now we'd know the assertion will always fail.
